### PR TITLE
Adding Windows support

### PR DESCRIPTION
**Pull request by op-ct**
This patch adds:
- New Windows facts:
  - Windows Server 2008 R2 and Windows Server 2012 R2 entries to the `Vagrantfile`
  - Windows Server 2008 R2 and Windows Server 2012 R2  facts for Facter 1.6–3.3
- New Scripts:
  - `facts/get_facts.ps1` to gather Windows facts
  - `facts/clean_facts.rb`, that cleans up `facts/` files (e.g., YAML-only output from early Windows Facter versions) before they are entered into the database.
  - A new Rake task (`rake table`) to generate a markdown matrix of OS versions x Facter versions currently supported by `facterdb` (to help maintain the `README.md`). 
- A new spec test to validate Windows facts
- An updated `README.md`

Note:
- `facts/clean_facts.rb` requires the Vagrant plugin [vagrant-host-shell](https://github.com/phinze/vagrant-host-shell)

### CODE REVIEW

**Comment by mcanevet** [comment]
@op-ct thanks a lot.
Wouldn't it be possible to convert UTF-16LE Unicode to UTF-8 or something similar so that we would have to do it at runtime?

**Comment by op-ct** [comment]
@mcanevet I'd rather have FacterDB be capable of handling UTF-16LE json files on its own.  It will allow folks to drop in `facter -j` dumps from their own Windows systems (especially if they didn't use the `Vagrantfile` scripts to collect them).

**Comment by mcanevet** [comment]
@op-ct well the problem is that I can't even see the .facts content as github detects it as Binary file

**Comment by mmoll** [comment]
what's the status here? I'd love to have this in so we can revert our workarounds in the theforeman-puppet tests for Windows. GH at least is viewing the file correctly with the "View" button in the diff.

**Comment by mcanevet** [comment]
@mmoll @op-ct I'd really prefer converting the UTF-16LTE to UTF-8 during facts collect.

**Comment by mmoll** [comment]
@op-ct in #22 some Windows versions got added, could you rebase and update your additionial WIndows facts to UTF-8? The CentOS updates could go into a seperate PR...

**Comment by wottah** [comment]
Are these issues getting resolved any time soon? Would greatly appreciate this fix.

File: facts/Vagrantfile [hunk 1]
```diff
@@ -91,12 +91,12 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
     host.vm.provision "shell", path: "get_facts.sh"
     host.vm.provision "shell", inline: "/sbin/shutdown -h now"
   end
-  config.vm.define "centos-7-x86_64" do |host|
-    host.vm.box = "puppetlabs/centos-7.0-64-nocm"
-    host.vm.provision "file", source: "Gemfile", destination: "Gemfile"
-    host.vm.provision "shell", path: "get_facts.sh"
-    host.vm.provision "shell", inline: "/sbin/shutdown -h now"
-  end
+ config.vm.define "centos-7-x86_64" do |host|
```
Comment (mcanevet) [inline facts/Vagrantfile]: Why did you change alignment?

**Comment by mcanevet** [comment]
@op-ct I still think the facts should be stored in plain text (UTF-8 or something)

File: facts/Vagrantfile (hunk 1 above)
Comment (op-ct) [inline facts/Vagrantfile]: That's some left-over silliness from separating the CentOS 7 fixes from the Windows; will be fixed in a bit.

**Comment by op-ct** [comment]
@mcanevet They should be stored in plain text after this (sanitized by `clean_facts.rb`).

**Comment by mcanevet** [comment]
@op-ct facts/3.1/windows-2008 R2-x86_64.facts is still in binary format

File: lib/facterdb.rb [hunk 2]
```diff
@@ -2,11 +2,30 @@
 require 'jgrep'
 
 module FacterDB
+  # Convenience method to handle odd files
```
Comment (mcanevet) [inline lib/facterdb.rb]: Is it still needed now that the facts are sanitized?

File: lib/facterdb.rb [hunk 3]
```diff
@@ -2,11 +2,23 @@
...
+        t = File.read(f)
+        [f,JSON.parse(t)]
+      end
+    ]
+  end
+
 
   def self.database
-    @database ||= "[#{Dir.glob("#{File.expand_path(File.join(File.dirname(__FILE__), '../facts'))}/*/*.facts").map { |f| File.read(f) }.join(',')}]\n"
+    fs = File.expand_path('../facts/*/*.facts',File.dirname(__FILE__))
```
Comment (mcanevet) [inline lib/facterdb.rb]: I don't get this change in the library...

Comment (op-ct) [inline lib/facterdb.rb]: It does a few things, but the intent was to be more semantic and less error-prone when a file contains bad data (or a different encoding).  

A convenient side effect is that `self.buld_database()` makes a nice place to insert code for inspecting incoming data.  I've already done this to identify broken data files during development.   In the future, we might consider using it to identify facter data that doesn't conform to FacterDB conventions (e.g., an `fqdn` of `foo.example.com`)

**Comment by DavidS** [comment]
Many windows versions were backfilled in https://github.com/camptocamp/facterdb/pull/46

@op-ct can you rebase and extract only the required code and docs changes you feel still applicable?

**Comment by glennsarti** [comment]
@op-ct I extracted your README change and put it into #76 .  Also enhanced the Windows detection support.

**Comment by rodjek** [comment]
Closing this out as there hasn't been any response and I believe that all the windows related improvements have been added in other PRs since this was opened.
